### PR TITLE
Add page on data availability

### PR DESCRIPTION
**Pull request by emmanuel-awosika**
This page discusses the concept of data availability, why it matters, and solutions to the data availability problem.

## Description

@samajammin suggested a page on data availability, so I decided to start things off with this contribution. The page breaks down the concept of data availability and how the data availability relates to security and scalability. It also has a brief description of different solutions to the data availability problem.

### CODE REVIEW

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:8]: I think we can explain this in simpler terms. The high-level ELI5 I'd expect to see here is that if we can’t reproduce something with the data we have available, from the blockchain’s perspective, it does not exist.

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:53]: I'd love to see an explicit section on monolithic vs modular blockchains. I think we imply it a lot but explaining how something like explaining monolithic blockchains do not have a data availability problem because every node replicates all data of every block, but in order to scale without increasing node requirements (decentralisation) we need to go modular. 

Not 100% sure where in this article that would best live, I'd leave that to your interpretation/vision.

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:10]: All developer docs should have a prerequisites section. Ideally we have this in mind when writing/reviewing as it informs the level of granularity we should go into about specific topics—each developer doc isn’t a stand-alone article, ideally, it should live relative to every other developer doc with clearly defined expected knowledge to comprehend.

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:2]: Meta: where do you see this living in the docs? Probably the advanced section? Inside or after directly after scaling?

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:145]: This stops quite abruptly based on previous content. I'd expect some sort of wrap-up section here. Perhaps pointing to the solution, Ethereum is leaning toward or the further work being done? Not sure but it just feels a little blunt of a stop to me.

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:12]: This feels like a very roundabout way of saying that all data should be available for transactions to reproduce the current world state

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:104]: Can we order the subheadings here from most to least impactful from the Ethereum ecosystem's perspective? In my mind data availability sampling is the favourite atm?

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:130]: Should we indicate that this is the solution we're leaning toward?

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:100]: We should be stronger in explaining the negative implications of off-chain DA here. 

If I read this as someone who didn't understand DA at all, these would feel more equally weighted in terms of potential solutions to the data availability problem than they actually are.

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:144]: Forgive my ignorance but isn't this a subsection of DAS? i.e. the current research suggests that DAS will include erasure coding and KZG commitment already?

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [reply src/content/developers/docs/data-availability.md:145]: Shouldn't we talk here about Danksharding?

**Review by minimalsm — CHANGES_REQUESTED** [review]
Hey @emmanuel-awosika, this is brilliant so far.

Done a big picture/high-level structural review, overall this is pretty solid, great work 😎. Some changes suggested here: 

- I'd like to see prerequisites added; 
- a little more depth on monolithic vs modular; 
- also had some questions about the solutions section, 
- and I think we should add some information on danksharding. 

Working on doing a copyediting review now.

File: src/content/developers/docs/data-availability.md
Comment (jmcook1186) [reply src/content/developers/docs/data-availability.md:8]: +1. I think it is also a nice place to include a "don't trust, verify" reminder, as ultimately this is why we care about data availability.

File: src/content/developers/docs/data-availability.md
Comment (jmcook1186) [inline src/content/developers/docs/data-availability.md:22]: This section kind of makes it seem like miners are not also full nodes. Besides, referring to miners puts a short expiry date ont he page since block production will be done by validators after the merge. Suggest we refer to "block producers" instead of miners/validators? This section could be simplified by replacing with something like:

"All nodes execute the transactions in blocks they receive from peers to ensure that the changes proposed by the block producer precisely match those computed independently by the node. This is how nodes verify that new information is valid, rather than having to trust that block producers are honest. This is not possible if any data is missing".

File: src/content/developers/docs/data-availability.md
Comment (jmcook1186) [inline src/content/developers/docs/data-availability.md:29]: I think an equally critical aspect is DA for zk and optimistic rollups

Comment (jmcook1186) [reply src/content/developers/docs/data-availability.md:29]: and also how data will be provably available on a stateless Ethereum (e.g. Portal network development)

File: src/content/developers/docs/data-availability.md
Comment (jmcook1186) [reply src/content/developers/docs/data-availability.md:100]: +1 It kind of undermines the "don't trust, verify" maxim that I suggest we repeat at the start of the intro

File: src/content/developers/docs/data-availability.md
Comment (jmcook1186) [reply src/content/developers/docs/data-availability.md:145]: Could be nice - just to say something like "state-of-the-art scaling proposals such as Danksharding rely upon DAS to verify that the entire contents of a blob have been seen by the network, relieving individual nodes from the burden of downloading and validating it all directly". I think there is some nuance to whether this translates into data retrievability - afaik it doesn't.

File: src/content/developers/docs/data-availability.md
Comment (minimalsm) [inline src/content/developers/docs/data-availability.md:110]: > Some validiums mitigate this problem by forcing block producers

I don't think mitigate is the correct phrasing here. (I'd personally argue that this strategy is wishful thinking in many cases)

**Review by minimalsm — CHANGES_REQUESTED** [review]
Copy editing run-through: never quite finished but thought I'd leave this here for now.

File: src/content/developers/docs/data-availability.md
Comment (jmcook1186) [reply src/content/developers/docs/data-availability.md:29]: AFAIK light clients will only download block headers unless they specifically request other data. Precisely how the data is then served to them isn't quite certain but likely to initially be via req/resp to a full node, then later by requesting from the p2p layer.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:2]: The advance section would be ideal. I also suggest place it after the page on scaling.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:130]: Based on @jmcook1186's suggestion, I added a new section to explain how Ethereum is approaching data availability (i.e., by using DAS).

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:144]: Yes, that's correct. I just made the data availability proofs an H4 to nest it under the DAS section.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md]: Added new text touching on both points. LMK what you think.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:8]: I created a new intro with ideas from both suggestions. Happy to hear your thoughts on it.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:10]: Added this now.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:104]: Fixed this.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:22]: I used your suggestion as the basis for the new intro, since it's the ELI5 explanation @minimalsm wanted to see in the intro. I did adjust the original text in this section to a simpler variant based on your suggestion.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:60]: > * Layer 2s execute off of Mainnet too
> That's what I said in the paragraph, or am I missing something there?

> > Usually compressed and posted on Ethereum in batches"
> 
> Usually? Are there cases where this isn't true for an L2?

No, it was a redundant word. Took it out now.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:110]: Changed the sentence to read "...validiums attempt to solve this problem..". And yes, DACs have risks, which I explained briefly in the intro to the following section (PoS-based DACs).

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:51]: Fixed this.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:53]: So, I weaved the distinction between modular and monolithic blockchains into the off-chain DA vs on-chain DA section. That sounded better than creating a new section for the topic.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md]: Added a new section explaining the relevance of the data availability to light clients.

Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md]: +1 on this. Most of the available material on DAS suggests that light clients will only have to download random chunks of a block's data to verify availability. The idea is that, with enough light clients sampling the same block, it's harder to conceal the unavailability of a block.

For context (from [Vitalik](https://hackmd.io/@vbuterin/sharding_proposal#ELI5-data-availability-sampling)): 
> Data availability sampling is in some ways a mirror image of randomly sampled committees. There is still sampling going on, in that each node only ends up downloading a small portion of the total data, but the sampling is done client-side, and within each blob rather than between blobs. Each node (including client nodes that are not participating in staking) checks every blob, but instead of downloading the whole blob, they privately select N random indices in the blob (eg. N = 20), and attempt to download the data at just those positions.

> The goal of this task is to verify that at least half of the data in each blob is available. If less than half of the data is available, then it is almost certain that at least one of the indices that any given client samples will not be available, and so the client will reject the blob.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:145]: I already included some info about sharding and data availability in the "decentralized scalability" section, where I also talk about layer 2s. But, since you requested, I moved the sharding + DAS part to the concluding section and added @jmcook1186's addition about Danksharding.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md]: Edited this part.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md]: > This seems a little simplistic. Obviously, Ethereum needs to allow data to be retrieved at some point past the block being confirmed.

Added a more specific description of data retrievability

Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md]: > > storing state data in perpetuity increases hardware requirements for full node
> 
> This technically isn't true. Full nodes are typically pruned, and requests are made to archive nodes to reconstruct the entirety of the chain.

Here are some Vitalik quotes for context on why I included that part:

> One of the longest and still unresolved challenges in the Ethereum protocol is that of how to manage the problem of growing state. Many actions in Ethereum (creating an account, filling a storage slot, sending ETH to a new account…) add objects to the state that all full nodes must store in order to be able to verify or produce new blocks. These actions require a one-time gas cost to be paid by the sender of the transaction, but impose permanent ongoing costs on the network, and the nodes that need to store this new data (and future nodes that need to download it when syncing).

> This is a major imbalance in the system’s design, and risks the Ethereum system becoming more and more difficult to use over time as the state fills with no-longer-needed “junk data”. The purpose of this document is to describe in more detail what the problem is, and what are some of the paths toward solving it. If we successfully implement a solution, it could pave the way to safely making significant increases to the block gas limit.

> The Ethereum state size is growing quickly. It is currently around 35 GB for just the state and over 100 GB including all Merkle proofs, and is increasing by roughly half that amount per year. State storage is additionally a weak point in Ethereum’s economics: it is the only mechanism by which a participant can pay a cost once to burden consensus nodes forever. In order to maintain the scalability and sustainability of Ethereum, we need some solution. Two paths to a solution exist, and have existed for a long time: weak statelessness and state expiry:

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:41]: > We only use the 'More on ....' style to link internally. Weakly in favour of moving these sections to further reading.

I'd have done that, but for these reasons:
- The further reading section already has many articles
- Some of the links point to relevant sections of each article, not the entire article (like the one on data retrievability).
- I felt linking to resources on each concept was more useful compared to throwing everything in the further reading section. The articles in the FR section tackle the overall concept of DA, not the individual solutions to the DA problem.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md:2]: Advanced section would make sense. Let's add it after scaling--I think DA is more than just scaling and needs its own page.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md]: Maybe you can explain this with more detail? Not totally sure what you mean here.

File: src/content/developers/docs/data-availability.md
Comment (emmanuel-awosika) [reply src/content/developers/docs/data-availability.md]: Fixed this.

**Review by minimalsm — APPROVED** [review]
Added to developer docs and merging 😀

Thanks @emmanuel-awosika and to everyone who helped review!

**Pull request merged**
